### PR TITLE
Implement skorch pruning callback

### PR DESCRIPTION
**Pull request by crcrpar**
Integration module for skorch and an example for it.

### CODE REVIEW

File: optuna/integration/skorch.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    Args:
+        trial:
+            A :class:`~optuna.trial.Trial` corresponding to the current evaluation of the
+            objective function.
+        monitor:
+            An evaluation metric for pruning, e.g. ``val_loss`` or
+            ``val_acc``. The metrics are obtained from the returned dictionaries,
+            i.e., ``net.histroy``. The names are thus depend on how this dictionary
+            is formatted.
```
Comment (HideakiImamura) [inline optuna/integration/skorch.py:25]: When ``val_loss`` is used, the user should specify `direction=minimize` in `optuna.create_study`. On the other hand, when ``val_acc`` is used, the user should specify ``direction=maximize` in `optuna.create_study`. I think it is kind for users to provide this information here.

File: optuna/integration/skorch.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+with optuna._imports.try_import() as _imports:
+    from skorch.callbacks import Callback
+    from skorch.net import NeuralNet
+
+if not _imports.is_successful():
+    Callback = object  # NOQA
+    NeuralNet = object  # NOQA
+
+
+class SkorchPruningCallback(Callback):
```
Comment (HideakiImamura) [inline optuna/integration/skorch.py:14]: How do we make this class "experimental" using the `experimental` decorator?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. This PR seems to be related to issue #1654. I have some early comments. 

I would like to know how this PR proceeds. What is your plan to remove the "draft"? I think adding an example is reasonable.

File: optuna/integration/skorch.py (hunk 2 above)
Comment (crcrpar) [reply optuna/integration/skorch.py:14]: What do you mean by that?

File: optuna/integration/skorch.py (hunk 1 above)
Comment (crcrpar) [reply optuna/integration/skorch.py:25]: That sounds a bit too much to me, and I don't think that kind comment should be on some other place like https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html?highlight=direction#optuna.study.Study.direction

**Comment by crcrpar** [comment]
Yeah, an example would be a required thing to mark this pull request as ready for review and that's my plan, thanks.

File: optuna/integration/skorch.py (hunk 1 above)
Comment (HideakiImamura) [reply optuna/integration/skorch.py:25]: I think the more helpful the documentation is, the better for users. I would like to know other reviewers' opinions.

File: optuna/integration/skorch.py (hunk 2 above)
Comment (HideakiImamura) [reply optuna/integration/skorch.py:14]: Sorry for my grammatical mistake. I mean: how about using the experimental decorator for this newly introduced class?
```suggestion
@experimental("2.1.0")
class SkorchPruningCallback(Callback):
```

Comment (crcrpar) [reply optuna/integration/skorch.py:14]: As per our internal discussion which is now published in https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#experimental-feature, I don't want to say this feature as "experimental".

File: optuna/integration/skorch.py (hunk 1 above)
Comment (crcrpar) [reply optuna/integration/skorch.py:25]: Okay, if that's the case, we should file some issues for some pruning callbacks that doesn't convey that information and label them "contribution-welcome".

File: optuna/integration/skorch.py (hunk 2 above)
Comment (HideakiImamura) [reply optuna/integration/skorch.py:14]: I got it. So this class is stable and there is no room for the interface to change at all? I think it's OK that we do not mark this class as "experimental".

By the way, how about other integration pruning callback classes? For example, I wonder whether the experimental decorator for the `CatalystCallback` can be removed.

File: optuna/integration/skorch.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    def __init__(self, trial: optuna.trial.Trial, monitor: str) -> None:
+
+        _imports.check()
+
+        super().__init__()
+        self._trial = trial
+        self._monitor = monitor
+
+    def on_epoch_end(self, net: NeuralNet, **kwargs: Any) -> None:
```
Comment (crcrpar) [inline optuna/integration/skorch.py:36]: this should be
```suggestion
    def on_epoch_end(self, net: "skorch.NeuralNet", **kwargs: Any) -> None:
```
?

**Comment by crcrpar** [comment]
@UTUnex Does this example look good to you? Any feedback is welcome 😃

**Comment by UTUnex** [comment]
Thank you all for your great efforts on the work. I checked the new example but I just find another module that is closer to what I expect for. The modul is __'optuna.integration.OptunaSearchCV ()'__, yeah, honestly, this is what I really like, which means I can just put all the required parameters in this function based on the demands.  I've just read the documentation about the optunasearchCV( ) and have some questions :  
(1) I don't know whether the estimator here can be the NeuralNetClassifer( ) in skorch, I haven't tried it but I hope it's compatible;
(2) I didn't find the option of tuning algorithm in the searchCV(), such as 'TPE', ''bayes', 'grid' or ' 'random'; if possible, I hope this can be added;
(3) Also, I didn't find the early stopping option in the searchCV( ), such as 'hyperband', 'BOHB' or the other advanced schemes which are useful for DL task tuning on large scale datasets;
(4) And, is it possible to add the single and multiple GPU support in the searchCV(), like searchCV( GPU usage = True, GPU_num = 2)
(5) How to output the searchCV results to a logger like tensorboard, ML flow, wandb etc, so that I can monitor and analyze the tuning process and hyperparameters combinations.

Summary: if possible, I hope you can add more functions to the optunasearchCV() and the expected code structure should be like : 
%% Data loading and preprocessing
     ....
%% Net design using pytorch  
     Class XXnet(nn.module)
         ....  
%% define the training function using skorch  
      net = NeuralNetClassifier( ...)  
%% hyperparameter tuning and logging to a certain dashboard:  
      search = optunasearchCV(net, **kwargs)
      search.fit(X_train, y_train)
%% Test  
      ...

The above is my advice, please check it. Thank you.

**Comment by crcrpar** [comment]
Thank you for your feedback!
Now that it seems that you are interested in OptunaSearchCV rather than this integration module which is almost out of scope of this pull request so that I (and maybe others) would like to support you in your original issue or gitter 😃

File: examples/skorch_simple.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
+"""
+Optuna example that optimizes multi-layer perceptrons using skorch.
+
+In this example, we optimize the validation accuracy of hand-written digit recognition using
+skorch, and MNIST. We optimize the neural network architecture. As it is too time
```
Comment (HideakiImamura) [inline examples/skorch_simple.py:5]: A nit.
```suggestion
skorch and MNIST. We optimize the neural network architecture. As it is too time
```

File: examples/skorch_simple.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+argument.
+    $ python skorch_simple.py [--pruning]
+
+"""
+
+import argparse
+
+import numpy as np
+import optuna
+from optuna.integration import SkorchPruningCallback
```
Comment (HideakiImamura) [inline examples/skorch_simple.py:18]: Import statements of `optuna` modules should be placed after the other third-party libraries like [this](https://github.com/optuna/optuna/blob/master/examples/pytorch_simple.py).

File: examples/skorch_simple.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+SUBSET_RATIO = 0.4
+
+mnist = fetch_openml("mnist_784", cache=False)
+
+X = mnist.data.astype("float32")
+y = mnist.target.astype("int64")
+indices = np.random.permutation(len(X))
+N = int(len(X) * SUBSET_RATIO)
+X = X[indices][:N]
+y = y[indices][:y]
```
Comment (HideakiImamura) [inline examples/skorch_simple.py:37]: I'm not sure, but
```suggestion
y = y[indices][:N]
```
?

File: examples/skorch_simple.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        callbacks=[SkorchPruningCallback(trial, "valid_acc")],
+    )
+
+    net.fit(X_train, y_train)
+
+    return accuracy_score(y_test, net.predict(X_test))
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="PyTorch Lightning example.")
```
Comment (HideakiImamura) [inline examples/skorch_simple.py:84]:
```suggestion
    parser = argparse.ArgumentParser(description="Skorch example.")
```

File: optuna/integration/skorch.py (hunk 3 above)
Comment (HideakiImamura) [reply optuna/integration/skorch.py:36]: Yes, I agree with you. As @hvy pointed in https://github.com/optuna/optuna/pull/1705#discussion_r476296682, this is rendered in the docs as `object`. Using `"skorch.NeuralNet"` instead `NeuralNet` leads to informative documentation.

**Review by HideakiImamura — COMMENTED** [review]
Thank you for your confirmation. I have several comments. Could you take a look?

**Review by HideakiImamura — APPROVED** [review]
Thank you for the update. LGTM!

File: optuna/integration/skorch.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+with optuna._imports.try_import() as _imports:
+    from skorch.callbacks import Callback
+    from skorch.net import NeuralNet
+
+if not _imports.is_successful():
+    Callback = object  # NOQA
+
+
+class SkorchPruningCallback(Callback):
```
Comment (crcrpar) [inline optuna/integration/skorch.py:13]: TODO: Add `versionadded` 

ref: https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#write-versionadded-directive-to-the-doc-of-new-features

File: optuna/integration/skorch.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    """Skorch callback to prune unpromising trials.
+
+    Args:
+        trial:
+            A :class:`~optuna.trial.Trial` corresponding to the current evaluation of the
+            objective function.
+        monitor:
+            An evaluation metric for pruning, e.g. ``val_loss`` or
+            ``val_acc``. The metrics are obtained from the returned dictionaries,
+            i.e., ``net.histroy``. The names are thus depend on how this dictionary
```
Comment (toshihikoyanase) [inline optuna/integration/skorch.py:23]: A nit. Or, it may be a typo for `The names are thus dependent on ...`.
```suggestion
            i.e., ``net.histroy``. The names thus depend on how this dictionary
```

File: tests/integration_tests/test_skorch.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    valid_ds = torch.utils.data.TensorDataset(torch.zeros(3, 4), torch.zeros(3, dtype=torch.long))
+
+    def objective(trial: optuna.trial.Trial) -> float:
+
+        net = skorch.NeuralNetClassifier(
+            ClassifierModule,
+            max_epochs=10,
+            lr=0.02,
+            callbacks=[SkorchPruningCallback(trial, "valid_acc")],
+            train_split=predefined_split(valid_ds),
```
Comment (toshihikoyanase) [inline tests/integration_tests/test_skorch.py:34]: In my environment, an error was occurred with `skorch==0.9.0`.  It works with `skorch==0.8.0`, and the new version may break compatibility.

**Review by toshihikoyanase — CHANGES_REQUESTED** [review]
Thank you for your PR. `skorch` seems to be developed actively, and I think the callback and example help its users. The change basically looks good to me, but I have some small comments. Please take a look.

File: optuna/integration/skorch.py (hunk 1 above)
Comment (crcrpar) [reply optuna/integration/skorch.py:24]: @toshihikoyanase what's your take here?

File: optuna/integration/skorch.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+    def on_epoch_end(self, net: "NeuralNet", **kwargs: Any) -> None:
+        history = net.history
+        if not history:
+            return
+        epoch = len(history) - 1
+        current_score = history[-1, self._monitor]
+        self._trial.report(current_score, epoch)
+        if self._trial.should_prune():
+            message = "Trial was pruned at epoch {}".format(epoch)
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply optuna/integration/skorch.py]: :+1:

File: tests/integration_tests/test_skorch.py (hunk 10 above)
Comment (crcrpar) [reply tests/integration_tests/test_skorch.py]: Fixed this in https://github.com/optuna/optuna/pull/1668/commits/ce004db0d68ff0190ae6e3f9d59b9e56e3bc32df.

File: optuna/integration/skorch.py (hunk 9 above)
Comment (crcrpar) [reply optuna/integration/skorch.py]: Thanks.

File: optuna/integration/skorch.py (hunk 1 above)
Comment (toshihikoyanase) [reply optuna/integration/skorch.py]: IMO, it is a bit verbose if we add the comments about `direction` here because the other pruning callbacks such as `KerasPruningCallback` and `XGBoostPruningCallback` also have the metrics similar to `monitor`.

So, I'd like to place such a comment in the reference of direction (`Study.direction` or `create_study`) or [tutorial](https://optuna.readthedocs.io/en/stable/tutorial/pruning.html#activating-pruners) of pruning.

**Review by toshihikoyanase — COMMENTED** [review]
The code and example look good to me.
Could you add `SkorchPruningCallback` reference to `docs/source/reference/integration.rst`?

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your update. LGTM!
I also confirmed that the CI job for the example worked in my local environment.
